x-pack/plugin/profiling/src/main/java/org/elasticsearch/xpack/profiling/action/TransportGetFlamegraphAction.java

```diff
@@ -73,22 +76,30 @@ static GetFlamegraphResponse buildFlamegraph(GetStackTracesResponse response) {
             return builder.build();
         }
 
-        SortedMap<String, StackTrace> sortedStacktraces = new TreeMap<>(response.getStackTraces());
-        for (Map.Entry<String, StackTrace> st : sortedStacktraces.entrySet()) {
-            StackTrace stackTrace = st.getValue();
-            builder.setCurrentNode(0);
+        Map<String, StackTrace> stackTraces = response.getStackTraces();
+        AtomicInteger lostStackTraces = new AtomicInteger();
+
+        response.getStackTraceEvents().forEach((eventId, event) -> {
+            StackTrace stackTrace = stackTraces.get(eventId.stacktraceID());
+            if (stackTrace == null) {
+                lostStackTraces.getAndIncrement();
+                return;
+            }
 
-            long samples = stackTrace.count;
-            builder.addSamplesInclusive(0, samples);
-            builder.addSamplesExclusive(0, 0L);
+            long samples = event.count;
+            double annualCO2Tons = event.annualCO2Tons;
+            double annualCostsUSD = event.annualCostsUSD;
 
-            double annualCO2Tons = stackTrace.annualCO2Tons;
-            builder.addAnnualCO2TonsInclusive(0, annualCO2Tons);
-            builder.addAnnualCO2TonsExclusive(0, 0.0d);
+            String executableName = eventId.executableName();
```

Comment:
Maybe rename these to avoid confusion later as it's actually `processName` we're storing:

```suggestion
            String processName = eventId.processName();
```

Reply:
Resolved